Tests/SwiftDiagnosticsTest/NewDiagnosticsFormatterTests.swift

```diff
@@ -0,0 +1,222 @@
+      ),
+      highlights: diagnosticDescriptor.highlight,
+      notes: notes,
+      fixIts: diagnosticDescriptor.fixIts
+    )
+    diagnostics.append(diagnostic)
+  }
+
+  let annotatedSource = DiagnosticsFormatter.annotatedSource(tree: tree, diags: diagnostics)
+
+  assertStringsEqualWithDiff(
+    annotatedSource,
+    expectedAnnotatedSource,
+    file: file,
+    line: line
+  )
+}
+
+final class DiagnosticsFormatterTests: XCTestCase {
+  func test_1() {
+    assertAnnotated(
+      markedSource: """
+      func foo() -> Int {
+        if 1️⃣1 != 0 {
+          return 0
+        }
+        return 1
+      }
+      """,
+      withDiagnostics: [
+        DiagnosticDescriptor(
+          locationMarker: "1️⃣",
+          id: .test,
+          message: "My message goes here!",
+          severity: .warning
+        ),
+      ],
+      matches: """
+        1 │ func foo() -> Int {
+        2 │   if 1 != 0 {
+          │      ╰─ warning: My message goes here!
+        3 │     return 0
+        4 │   }
+
+        """
+    )
+  }
+
+  func test_2() {
+    assertAnnotated(
+      markedSource: """
+      func foo() -> Int {
+        if 1️⃣1 != 0 2️⃣{
+          return 0
+        }
+        return 1
+      }
+      """,
+      withDiagnostics: [
+        DiagnosticDescriptor(
+          locationMarker: "1️⃣",
+          id: .test,
+          message: "My message goes here!",
+          noteDescriptors: [
+            NoteDescriptor(locationMarker: "1️⃣", id: .test, message: "First message"),
+            NoteDescriptor(locationMarker: "1️⃣", id: .test, message: "Second message"),
+            NoteDescriptor(locationMarker: "2️⃣", id: .test, message: "Other message")
+          ]
+        ),
+      ],
+      matches: """
+        1 │ func foo() -> Int {
+        2 │   if 1 != 0 {
+          │      │      ╰─ note: Other message
+          │      ├─ error: My message goes here!
+          │      ├─ note: First message
+          │      ╰─ note: Second message
+        3 │     return 0
+        4 │   }
+
+        """
+    )
+  }
+
+  func test_3() {
+    assertAnnotated(
+      markedSource: """
+      func 2️⃣foo() -> Int {
+        if 1️⃣1 != 0 {
+          return 0
+        }
+        return 1
+      }
+      """,
+      withDiagnostics: [
+        DiagnosticDescriptor(
+          locationMarker: "1️⃣",
+          id: .test,
+          message: "My message goes here!",
+          noteDescriptors: [
+            NoteDescriptor(locationMarker: "1️⃣", id: .test, message: "First message"),
+            NoteDescriptor(locationMarker: "2️⃣", id: .test, message: "Second message"),
+          ]
+        ),
+      ],
+      matches: """
+        1 │ func foo() -> Int {
+          |      ╰─ note: Second message // This note comes from from diagnostic, how to present it?
+        2 │   if 1 != 0 {
+          │      ├─ error: My message goes here!
+          │      ╰─ note: First message
+        3 │     return 0
+        4 │   }
```

Comment:
I think this looks good if we use `GroupedDiagnosticFormatter`. The idea would be that each error basically gets its own printout of the source (we wouldn’t be merging multiple diagnostics into a single source printout anymore) and thus it’s clear which diagnostic the note belongs to.

Reply:
Yes, this looks good to me, too. Right now, the Swift compiler is intentionally grouping things so that there is only a single top-level diagnostic, because that made sure all of the information about a single issue (with all of its notes) would occur in the one printout.

Notes in different files are still a little tricky---they look like top-level things right now, but @ahoppen has reasonably suggested that they should be in a "box" as if they were nested, because notes are never standalone---they're always associated with some other error/warning/remark.